### PR TITLE
PWX-24444: Antihyperconvergence scheduling of sharedv4 service pods

### PR DESCRIPTION
**Pull request by pp511**
Signed-off-by: Priyanshu Pandey <ppandey@purestorage.com>

**What type of PR is this?**
>feature
>unit-test

**What this PR does / why we need it**:
- Introduce a new parameter stork.libopenstorage.org/preferRemoteNodeOnly parameter to be applied to StorageClass. 
Pods using this StorageClass should only be scheduled on the nodes where volume replica nodes do not exist. This behavior should be strictly enforced.
- For Pods using SharedV4 Service  volumes, Stork should give higher score to nodes where SharedV4 Service volumes do not exist.
- Refer to PWX-24444 for further details

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:
```release-note
*  Stork will try to schedule application pods using SharedV4 service volumes to run on nodes where volume replica does not exist such that it uses NFS mountpoint and when a failover happens application pods will not be restarted.
* StorageClass parameter "stork.libopenstorage.org/preferRemoteNodeOnly": "true" can be used to strictly enforce this behavior.


```

**Does this change need to be cherry-picked to a release branch?**:
yes 2.12.2

**Notes For the Reviewer**
- Pod annotation preferLocalNodeOnly doesn't apply for SharedV4 service volumes
- StorageClass parameter preferRemoteNodeOnly applies only to SharedV4 service volumes
- In case a Pod uses both SharedV4 service and regular volumes, scoring will prioritize antihyperconvergence based on sharedv4 volume replicas

### CODE REVIEW

File: pkg/extender/extender.go [hunk 1]
```diff
@@ -315,6 +342,19 @@ func (e *Extender) processFilterRequest(w http.ResponseWriter, req *http.Request
...
 
+// volumePrefersRemoteOnly checks if preferRemoteNodeOnly label is applied to the label
+func (e *Extender) volumePrefersRemoteOnly(volumeInfo *volume.Info) bool {
+	if volumeInfo.Labels != nil {
+		if value, ok := volumeInfo.Labels[preferRemoteNodeOnlyParameter]; ok {
+			if preferRemoteOnlyExists, err := strconv.ParseBool(value); err == nil {
+				return preferRemoteOnlyExists
+			}
+		}
+		return false
```
Comment (diptiranjanpx) [inline pkg/extender/extender.go:353]: redundant

File: pkg/extender/extender_test.go [hunk 2]
```diff
@@ -1534,3 +1546,476 @@ func extenderMetricsTest(t *testing.T) {
...
+			defaultScore,
+			defaultScore,
+			defaultScore,
+			defaultScore,
+			defaultScore},
+		prioritizeResponse)
+}
+
+// Verify disableHyperconvergenceAnnotation honors sharedv4 service volumes
+func multiVolumeDisableHyperconvergenceTest(t *testing.T) {
```
Comment (adityadani) [inline pkg/extender/extender_test.go:1872]: nit: this test is not multi-volume

File: pkg/extender/extender_test.go [hunk 3]
```diff
@@ -1534,3 +1546,476 @@ func extenderMetricsTest(t *testing.T) {
...
+	filterResponse, err := sendFilterRequest(pod, nodes)
+	if err != nil {
+		t.Fatalf("Error sending filter request: %v", err)
+	}
+	verifyFilterResponse(t, nodes, []int{1, 2, 4, 5}, filterResponse)
+}
+
+// Deploy both sharedv4 service volume and regular volumes
+// Verify nodes where pods using sharedv4 service volumes can be antihyperconverged takes higher score
+func multiVolumeAntihyperconvergenceTest(t *testing.T) {
```
Comment (adityadani) [inline pkg/extender/extender_test.go:1729]: Can you add a multiVolume test where there are 6 nodes and the volume replicas are distributed in the following way:

Case 1:
- regularVolume: 0, 1, 2
- sharedv4SvcVolume: 2, 3, 4

priorityResponse: {2*nodePriority, 2*nodePriority, defaultScore, defaultScore, defaultScore}

Case 2: 

- regularVolume: 0, 1, 2
- sharedv4SvcVolume: 0, 1, 2

filterRequest should return {0, 1, 2}
prioritizeRequest should return {defaultScore, defaultScore, defaultScore}

**Review by adityadani — APPROVED** [review]
lgtm! Really good set of UTs. 
Please add the release note once you are ready to merge

File: pkg/extender/extender_test.go (hunk 3 above)
Comment (pp511) [reply pkg/extender/extender_test.go:1729]: Case2: 
With preferRemoteOnlyNode enabled, filter will return (3,4, 5) in filter response
Without preferRemoteOnly, filter will return all nodes

File: pkg/extender/extender.go [hunk 4]
```diff
@@ -251,12 +263,20 @@ func (e *Extender) processFilterRequest(w http.ResponseWriter, req *http.Request
...
 
-			nodeVolumeCounts := make(map[string]int)
-			if preferLocalOnly {
-				// Get nodes that have replicas for all the volumes
-				for _, volumeInfo := range driverVolumes {
-					for _, volumeNode := range volumeInfo.DataNodes {
-						nodeVolumeCounts[volumeNode]++
+			nonSharedV4SvcVolumeCount := 0
+			nodeVolumeCountWithLocalReplica := make(map[string]int)
+			nodeSharedV4SvcVolumeExists := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (pp511) [reply pkg/extender/extender.go]: Removed all sharedv4-svc references from variable names, test case names and comments

File: pkg/extender/extender_test.go (hunk 2 above)
Comment (pp511) [reply pkg/extender/extender_test.go]: Ack. Renamed the test

File: pkg/extender/extender_test.go (hunk 3 above)
Comment (pp511) [reply pkg/extender/extender_test.go]: Added two more test cases. Handled case 2 with preferRemoteNodeOnly as I commented above

File: pkg/extender/extender.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pp511) [reply pkg/extender/extender.go]: I handled it differently. Not using this variable anymore. Removing this

**Pull request merged**